NuLink M1&M2

pawnz0 opened the pull request:
# Milestone Delivery Checklist

- [x] The [milestone-delivery-template.md](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/deliveries/milestone-delivery-template.md) has been copied and updated.
- [x] The [invoice form :pencil:](https://forms.gle/LSRr7PCjBpEbKGh89) has been filled out for this milestone.
- [x] This pull request is being made by the same account as the accepted application.
- [x] In case of acceptance, the payment will be transferred to the BTC/ETH payment address provided in the application.
- [x] The delivery is according to the [Guidelines for Milestone Deliverables](https://github.com/w3f/Grants-Program/blob/master/docs/milestone-deliverables-guidelines.md).

Link to the application pull request: https://github.com/w3f/Grants-Program/pull/455 < please fill this in with the PR number of your application.

alxs commented:
Thanks for the delivery @pawnz0! We'll look into it as soon as possible.

alxs commented on deliveries/nulink_milestone_1.md:
```diff
@@ -0,0 +1,22 @@
+**The [invoice form :pencil:](https://docs.google.com/forms/d/e/1FAIpQLSfmNYaoCgrxyhzgoKQ0ynQvnNRoTmgApz9NrMp-hd8mhIiO0A/viewform) has been filled out correctly for this milestone and the delivery is according to the official [milestone delivery guidelines](https://github.com/w3f/General-Grants-Program/blob/master/grants/milestone-deliverables-guidelines.md).**
+
+
+* **PR Link:** https://github.com/w3f/Grants-Program/pull/455
```
```suggestion
* **Application Document:** https://github.com/w3f/Grants-Program/blob/master/applications/NuLink.md
```

alxs commented:
Hey @pawnz0. Sorry for the late reply. I finally had a chance to have a look at your delivery today and have the following initial comments:

- The project currently fails to build due to misusing inner doc comments - [this commit](https://github.com/NuLink-network/nulink-chain/commit/a228626ec869e0b878caac13dd83440a9d7d359d) breaks things. Please check if it builds for you and that you haven't disabled any additional lints. Ideally, you could set up CI on your repository.
- One test is currently failing:
  ```
  failures:
  
  ---- tests::it_works_for_policy stdout ----
  thread 'tests::it_works_for_policy' panicked at 'assertion failed: `(left == right)`
    left: `195`,
   right: `333`', pallets/nuproxy/src/tests.rs:310:9
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
  ```
- Further, a number of tests are currently commented out. Why is this?
- `cargo clippy` returns a number of warnings in both pallets and the node, please address these.
- You proposed to use a threshold signature-like scheme for deliverables 2 `ValidateWatchers` of milestone 1; however, the current implementation simply accepts anything coming from a watcher. Seeing how this was also a [point of discussion](https://github.com/w3f/Grants-Program/pull/455#issuecomment-872221432) during your application, I'd like to ask you to please submit an [amendment](https://github.com/w3f/Grants-Program#changes-to-a-grant-after-approval) to record any differences between your proposal and the current delivery.

pawnz0 commented:
Thanks for your review.

1. We have fixed the issue and it should be ready for build now.
2.  The test have been fixed.
3.  Those are some test we write when developing. We forget to delete it before submit. And it is now clear.
4.  Most of the warning has been addressed and fixed. 
5.  We have submitted the amendment [here](https://github.com/w3f/Grants-Program/pull/846).

alxs commented:
On hold pending approval of the amendment.

pawnz0 commented:
Hello, Alex. We have made some updates as we discuss:

1, We have added detail instruction of the method [here](https://github.com/NuLink-network/nulink-chain/blob/main/README.md#usage).
2, You can check how watcher nodes works [here](https://github.com/NuLink-network/nulink-watcher/blob/main/README.md#work-flow). And remark that we provide the update extrinsics in the pallet but the watcher would automatically deal with this update. 
3. We have put the watcher configuration file in the project tree as you suggest and the developer could edit [here](https://github.com/NuLink-network/nulink-watcher/blob/main/config.json) directly if needed.
4. The bonding workers and stakers are the same people in the NuCypher network, each worker needs to bond to a staker address before providing service and gaining reward.

pawnz0 commented:
Hello, please check the updated [instruction ](https://github.com/NuLink-network/nulink-chain/blob/main/README.md#usage)and let me know if you get any further questions.

alxs commented:
Thanks @pawnz0. A few follow-up questions:

- `get_active_staker` doesn't do random sampling as initially described. Why is this and what was the intention in the first place?
- Could you please point me to the NuCypher documentation on how rewards are calculated, in order to better understand your implementation?
- `cargo clippy` again returns multiple warnings, please address them.
- Could you please provide complete instructions to create a policy? Or link to the NuCypher documentation to create one and explain that one can then register it on NuLink.

alxs commented:
Thanks. That makes sense.

Thanks also for the explanation regarding the distribution of rewards. If the current scheme has been discussed with the NuCypher team, that's fine by me.
 
Great, that's what I was looking for regarding policy creation. You may want to link to that document [here](https://github.com/NuLink-network/nulink-chain#create-policy-by-user) for more info instead of to the code.

Could you please update the links in your deliveries, since the line numbers are now off? Please also include a commit hash in the links so they will continue to work in the future. You can automatically turn URLs into permalinks by pressing `y`on GitHub.

I can't run your watcher - below is the output. I can connect to the chain on the same port from the apps portal.

```
INFO [03-23|11:04:05.475] Connecting to substrate chain...         url=ws://127.0.0.1:9944
2022/03/23 11:04:05 Connecting to ws://127.0.0.1:9944...
INFO [03-23|11:04:05.562] Submitting substrate call...             method=NulinkNuproxy.register_watcher sender=5EZQrJ2XBSNnzQ78jEBiRZj5S4SMEczZJMc8ANL54AWdtz9n
ERROR[03-23|11:04:05.590] failed to register watcher               error="failed to construct call, err: module NulinkNuproxy not found in metadata for call NulinkNuproxy.register_watcher"
ERROR[03-23|11:04:05.590] run failed                               error="failed to construct call, err: module NulinkNuproxy not found in metadata for call NulinkNuproxy.register_watcher"

```

By the way, it seems your endpoint is down.

pawnz0 commented:
@alxs  Hello. Thanks for reviewing. 

1. We have put the details of creating a policy link in the instruction.
2. We have changed the link to the permalink as you suggest.
3. Please update  both pallet and watcher to the newest version. And check that the config in watcher is same as the image below. 
![config](https://user-images.githubusercontent.com/53824975/159865743-55c1aa40-c4d6-4800-8996-4cbf5c0aa7be.jpg)

And then you can start running your local node 
![node](https://user-images.githubusercontent.com/53824975/159865724-99bee401-1f69-42e8-9f87-83bfe01e1ff1.jpg)

After success running, you can now start your watcher :
![watcher](https://user-images.githubusercontent.com/53824975/159865733-b595fbc6-5ab8-4253-978e-1ffed687831b.jpg)

4. The endpoint is now active.

alxs commented:
Thanks. That is still not working for me (same error as before). Could you try on a fresh install or at least with a freshly cloned repo?

alxs commented:
@pawnz0 seems the problem was indeed on my end. Thanks for reaching out. You may want to provide a more fitting error message when the watcher can't connect to the chain.

With this, I'm happy to tell you that the milestone is a pass  - you can find my evaluation notes [here](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/evaluations/nulink_1_alxs.md) and [here](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/evaluations/nulink_2_alxs.md). I'll notify the operations team to pay your invoice. Please allow for up to 14 days for your payment to be processed. Best of luck with the continuation of your project!

fededubbi commented:
Good morning,

I noticed that on both invoices you mentioned amount BTC, but the actual amount has DAI as currency.
Can you please delete BTC if we need to pay in DAI and resend us the invoices?

Many thanks,
regards,

Federica

pawnz0 commented:
> Good morning,
> 
> 
> 
> I noticed that on both invoices you mentioned amount BTC, but the actual amount has DAI as currency.
> 
> Can you please delete BTC if we need to pay in DAI and resend us the invoices?
> 
> 
> 
> Many thanks,
> 
> regards,
> 
> 
> 
> Federica 

Np, how could I resend them? The same link or email you?

fededubbi commented:
Good morning,

You can send them directly to this email address: invoices@web3.foundation.

Many thanks,
Regards,

Federica Dubbini

pawnz0 commented:
> Good morning,
> 
> 
> 
> You can send them directly to this email address: invoices@web3.foundation.
> 
> 
> 
> Many thanks,
> 
> Regards,
> 
> 
> 
> Federica Dubbini

Sent, please check! Thanks.
